pkg/service.go

```diff
@@ -186,51 +196,58 @@ func (s *Service) knownValidatorsUpdateTime() time.Time {
 	return updateTime
 }
 
-func (s *Service) updateProposerDuties(ctx context.Context, client BeaconClient, headSlot structs.Slot) error {
+func (s *Service) getProposerDuties(ctx context.Context, client BeaconClient, headSlot structs.Slot) (entries []RegisteredProposersResponseData, err error) {
 	epoch := headSlot.Epoch()
 
-	logger := s.Log.With(log.F{
-		"method":    "UpdateProposerDuties",
-		"slot":      headSlot,
-		"epochFrom": epoch,
-		"epochTo":   epoch + 1,
-	})
-
-	state := structs.DutiesState{}
-
 	// Query current epoch
 	current, err := client.GetProposerDuties(epoch)
 	if err != nil {
-		return fmt.Errorf("current epoch: get proposer duties: %w", err)
+		return nil, fmt.Errorf("current epoch: get proposer duties: %w", err)
 	}
 
-	entries := current.Data
+	entries = current.Data
 
 	// Query next epoch
 	next, err := client.GetProposerDuties(epoch + 1)
 	if err != nil {
-		return fmt.Errorf("next epoch: get proposer duties: %w", err)
+		return nil, fmt.Errorf("next epoch: get proposer duties: %w", err)
 	}
-	entries = append(entries, next.Data...)
 
-	state.ProposerDutiesResponse = make(structs.BuilderGetValidatorsResponseEntrySlice, 0, len(entries))
-	state.CurrentSlot = headSlot
+	return append(entries, next.Data...), nil
+}
+
+func (s *Service) storeProposerDuties(ctx context.Context, d Datastore, vCache ValidatorCache, headSlot structs.Slot, entries []RegisteredProposersResponseData) {
+	epoch := headSlot.Epoch()
+	logger := s.Log.With(log.F{
+		"method":    "UpdateProposerDuties",
+		"slot":      headSlot,
+		"epochFrom": epoch,
+		"epochTo":   epoch + 1,
+	})
+
+	state := structs.DutiesState{
+		CurrentSlot:            headSlot,
+		ProposerDutiesResponse: make(structs.BuilderGetValidatorsResponseEntrySlice, 0, len(entries)),
+	}
 
+	var err error
 	for _, e := range entries {
-		reg, err := s.Datastore.GetRegistration(ctx, e.PubKey)
-		if err == nil {
-			state.ProposerDutiesResponse = append(state.ProposerDutiesResponse, types.BuilderGetValidatorsResponseEntry{
-				Slot:  e.Slot,
-				Entry: &reg,
-			})
-		} else if err != nil && !errors.Is(err, ds.ErrNotFound) {
-			logger.Warn(err)
+		reg, ok := vCache.Get(e.PubKey.PublicKey)
+		if !ok {
+			reg.Entry, err = d.GetRegistration(ctx, e.PubKey.PublicKey)
+			if err != nil {
+				if !errors.Is(err, datastore.ErrNotFound) {
+					logger.Warn(err)
```

Comment:
Could we add something like. `logger.Warn(fmt.Errorf("fail to register validator %s: %w", pubkey, err))`?

Reply:
as you wish :)